src/cryptography/hazmat/backends/openssl/aead.py

```diff
@@ -22,11 +22,216 @@
         AESCCM, AESGCM, AESOCB3, AESSIV, ChaCha20Poly1305
     ]
 
+
+def _is_evp_aead_supported_cipher(
+    backend: Backend, cipher: _AEADTypes
+) -> bool:
+    """Check if cipher is ChaCha20Poly1305 and backend supports EVP_AEAD
+    (i.e., is BoringSSL or LibreSSL).
+
+    This enables the EVP_AEAD codepath for backends that support it.
+    """
+    from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
+
+    boring_or_libre = (
+        backend._lib.CRYPTOGRAPHY_IS_BORINGSSL
+        or backend._lib.CRYPTOGRAPHY_IS_LIBRESSL
+    )
+    return boring_or_libre and isinstance(cipher, ChaCha20Poly1305)
+
+
+def _aead_cipher_supported(backend: Backend, cipher: _AEADTypes) -> bool:
+    if _is_evp_aead_supported_cipher(backend, cipher):
+        return True
+    else:
+        cipher_name = _openssl_aead_cipher_name(cipher)
+        if backend._fips_enabled and cipher_name not in backend._fips_aead:
+            return False
+        # SIV isn't loaded through get_cipherbyname but instead a new fetch API
+        # only available in 3.0+. But if we know we're on 3.0+ then we know
+        # it's supported.
+        if cipher_name.endswith(b"-siv"):
+            return backend._lib.CRYPTOGRAPHY_OPENSSL_300_OR_GREATER == 1
+        else:
+            return (
+                backend._lib.EVP_get_cipherbyname(cipher_name)
+                != backend._ffi.NULL
+            )
+
+
+def _aead_create_ctx(
+    backend: Backend,
+    cipher: _AEADTypes,
+    key: bytes,
+):
+    if _is_evp_aead_supported_cipher(backend, cipher):
+        return _evp_aead_create_ctx(backend, cipher, key)
+    else:
+        return _openssl_aead_create_ctx(backend, cipher, key)
+
+
+def _encrypt(
+    backend: Backend,
+    cipher: _AEADTypes,
+    nonce: bytes,
+    data: bytes,
+    associated_data: typing.List[bytes],
+    tag_length: int,
+    ctx: typing.Any = None,
+) -> bytes:
+    if _is_evp_aead_supported_cipher(backend, cipher):
+        return _evp_aead_encrypt(
+            backend, cipher, nonce, data, associated_data, tag_length, ctx
+        )
+    else:
+        return _openssl_encrypt(
+            backend, cipher, nonce, data, associated_data, tag_length, ctx
+        )
+
+
+def _decrypt(
+    backend: Backend,
+    cipher: _AEADTypes,
+    nonce: bytes,
+    data: bytes,
+    associated_data: typing.List[bytes],
+    tag_length: int,
+    ctx: typing.Any = None,
+) -> bytes:
+    if _is_evp_aead_supported_cipher(backend, cipher):
+        return _evp_aead_decrypt(
+            backend, cipher, nonce, data, associated_data, tag_length, ctx
+        )
+    else:
+        return _openssl_decrypt(
+            backend, cipher, nonce, data, associated_data, tag_length, ctx
+        )
+
+
+# EVP_AEAD support follows.
```

Comment:
NB: I've added these comments for myself (to visually separate the internal impls), but can remove.